### ISSUE TITLE
[ENH] in `TimesFMForecaster`, `context_len` and `horizon_len` should be set automatically per default

### ISSUE THREAD

**Issue by fkiraly** [sktime/sktime#7209]
`TimesFMForecaster` should set `context_len` and `horizon` len to a reasonable value, automatically per default.

The `horizon_len` can be obtained from `fh`, but it makes sense to give the user the option to override.

The `context_len` can also be set to an automatic choice, in the default. Again, giving the user the option to override makes sense.

FYI @geetu040.

As a recipe to change this, there needs to be an internal `self._context_len` and `self._horizon_len` which is determined by the settings.

**Comment by geetu040** [comment]
yes that sure sounds good. there are no specific values ideal for the defaults, but there are a few considerations
- `context_length` should be a multiple of `input_patch_len`, which by default is `input_patch_len=32` for loading the model checkpoint.
- `horizon_len` should ideally be a less than or equal to `context_len`

In the [latest code](https://github.com/google-research/timesfm/blob/master/src/timesfm/timesfm_base.py#L79), the authors have by default kept these values to be
```py
context_len = 512
horizon_len = 128
```

**Comment by fkiraly** [comment]
Is there an algorithmic difference in the output, from wetting `horizon_len` to 42 and then taking the first 12 values, or setting `horizon_len` to 12 in the first place?

**Comment by geetu040** [comment]
> Is there an algorithmic difference in the output, from wetting `horizon_len` to 42 and then taking the first 12 values, or setting `horizon_len` to 12 in the first place?

No, there is no difference in the output
It produces forecasts for any horizon using auto-regressive decoding

**Comment by fkiraly** [comment]
then why should we not set it to max `fh` always, at least in the default?

**Comment by avishkarsonni** [comment]
Can I pick this issue up?

**Comment by avishkarsonni** [comment]
Can I work on this issue?

**Comment by fkiraly** [comment]
sure, just make sure you have agreed on clear defaults for the parameters before you implement them. What would you suggest?

**Comment by shivamgcodes** [comment]
Hey, I am interested in working on this issue ..

Setting the default value of horizon_len to max fh sounds good to me,
Or I can set it to whatever else you suggest ..

**Comment by fkiraly** [comment]
@geetu040, does default `horizon_len` = max-fh sound reasonable?

Also, what should `context_len` be? We can work on that separately of course.

**Comment by geetu040** [comment]
> @geetu040, does default `horizon_len` = max-fh sound reasonable?

I designed it the way it is, because this allows the model to be flexible towards different `fh` during predict. If any of the `fh` is greator than `horizon_len` during fit, `horizon_len` is overridden and during predict, exception is raised.

Now by default `horizon_len=None` and interpret it from `fh` makes sense, but I would prefer that `max(fh)` overrides `horizon_len` only when `horizon_len=None` (new condition) or `horizon_len < max(fh)` (already there).

> Also, what should `context_len` be? We can work on that separately of course.

I am not too sure about `context_len`, it has be to `>=horizon_len`, we can set that automatically from the horizon len, that is not a bad approach, but I would prefer to take it from the user instead of setting that automatically because there is not good standard to creating context_len from horizon_len.

**Comment by fkiraly** [comment]
> Now by default `horizon_len=None` and interpret it from `fh` makes sense, but I would prefer that `max(fh)` overrides `horizon_len` only when `horizon_len=None` (new condition) or `horizon_len < max(fh)` (already there).

Makes sense, I would agree with that.

> I am not too sure about `context_len`, it has be to `>=horizon_len`, we can set that automatically from the horizon len, that is not a bad approach, but I would prefer to take it from the user instead of setting that automatically because there is not good standard to creating context_len from horizon_len.

What about setting it equal to `horizon_len` by default, or a constant multiple of it?

**Comment by geetu040** [comment]
> What about setting it equal to `horizon_len` by default, or a constant multiple of it?

yes we can do something like this, ensuring it meets these requirements
1. `context_len` should be a multiple of `self.input_patch_len` (raises exception otherwise, strict check)
2. `context_len` should be `>= horizon_len` (suggested, not a strict check)
3. generally the greater the `context_len` is, the better are the outputs. Ideally `context_len` should be as nearest as possible to `len(y)` following the first 2 points

**Comment by fkiraly** [comment]
How about this then: we round `context_len` up to the closest multiple of `self.input_patch_len`?

**Comment by geetu040** [comment]
> How about this then: we round `len(y)` up to the closest multiple of `self.input_patch_len`, and take that value as `context_len`?

sounds good to me

**Comment by fkiraly** [comment]
Ok, to summarize for @shivamgcodes, we agreed that:

* default `horizon_len` is `max(fh)`
* default `context_len` is the smallest multiple of `input_patch_len` that exceeds `len(y)`

The "calculated" variables should be set to `self.horizon_len_` and `context_len_` in `_fit`, in the default case, and in case they are explicitly set, these variables are still written to `self` but are equal to what the user sets.

**Comment by shivamgcodes** [comment]
sure, i will get to it.
Thanks.

**Comment by tanvincible** [comment]
Hi, @shivamgcodes! Are you working on this issue?

**Comment by shivamgcodes** [comment]
no, i am not working on it, you can take it.
